File: cafe-occidente-frontend/src/app/core/services/green-coffee-purchase.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { API_BASE_URL } from '../config/api.config';
import {
  GreenAnnouncementInfo,
  GreenCoffeePurchaseCalculation,
  GreenCoffeePurchaseRequest,
  GreenCoffeePurchaseResponse,
  GreenNextInvoiceNumber,
} from '../models/green-coffee-purchase.model';

/** Unica responsabilidad: llamadas HTTP para el formulario Compras Cafe Verde (VERDES). */
@Injectable({ providedIn: 'root' })
export class GreenCoffeePurchaseService {
  private readonly http = inject(HttpClient);

  create(request: GreenCoffeePurchaseRequest): Observable<GreenCoffeePurchaseResponse> {
    return this.http.post<GreenCoffeePurchaseResponse>(`${API_BASE_URL}/purchases/green-coffee`, request);
  }

  preview(request: GreenCoffeePurchaseRequest): Observable<GreenCoffeePurchaseCalculation> {
    return this.http.post<GreenCoffeePurchaseCalculation>(
      `${API_BASE_URL}/purchases/green-coffee/preview`,
      request,
    );
  }

  /** Factura: siguiente consecutivo propio de VERDES. */
  nextInvoiceNumber(): Observable<GreenNextInvoiceNumber> {
    return this.http.get<GreenNextInvoiceNumber>(`${API_BASE_URL}/purchases/green-coffee/next-invoice-number`);
  }

  /** Cod Prod + anuncio vigente (Especial "CV" / Fondo "RP" fijos). */
  announcementInfo(): Observable<GreenAnnouncementInfo> {
    return this.http.get<GreenAnnouncementInfo>(`${API_BASE_URL}/purchases/green-coffee/announcement-info`);
  }
}

File: cafeoccidente-backend/src/main/java/com/cafeoccidente/backend/purchases/greencoffee/service/GreenCoffeePurchaseService.java
package com.cafeoccidente.backend.purchases.greencoffee.service;

import com.cafeoccidente.backend.purchases.greencoffee.dto.AnnouncementInfoResponse;
import com.cafeoccidente.backend.purchases.greencoffee.dto.GreenCoffeePurchaseRequest;
import com.cafeoccidente.backend.purchases.greencoffee.dto.GreenCoffeePurchaseResponse;
import com.cafeoccidente.backend.purchases.greencoffee.dto.NextInvoiceNumberResponse;

public interface GreenCoffeePurchaseService {
    GreenCoffeePurchaseResponse create(GreenCoffeePurchaseRequest request);

    GreenCoffeePurchaseResponse findById(Long id);

    /** Recalcula la cascada completa sin persistir. */
    GreenCoffeePurchaseCalculation preview(GreenCoffeePurchaseRequest request);

    /** Factura: siguiente consecutivo propio de VERDES. */
    NextInvoiceNumberResponse nextInvoiceNumber();

    /** Cod Prod + anuncio vigente para la agencia del usuario (Fondo "RP" / Especial "CV" fijos). */
    AnnouncementInfoResponse announcementInfo();
}
